Episode 2 (#67)
carpentries-incubator/deep-learning-intro: _episodes/02-keras.md

~~~
{:.language-python}

Second, the target data is also in a format that cannot be used to train.
To train a neural network we need to be able to calculate how "far away" the species predicted by the
neural network is from the true species.
When the target is a string category column as we have here it is very difficult to determine this "distance" or error.
Therefore we will transform this column into a more suitable format.
Again there are many ways to do this, however we will be using the 1-hot encoding.
This encoding creates multiple columns, as many as there are unique values, and
puts a 1 in the column with the corresponding correct class, and 0's in
the other columns.
For instance, for a penguin of the Adelie species the 1 hot encoding would be 1 0 0

Fortunately pandas is able to generate this encoding for us.
~~~
target = pd.get_dummies(valid_data['species'])
target.head() # print out the top 5 to see what it looks like.
~~~
{:.language-python}

> ## Keras for Neural Networks
> For this lesson we will be using [Keras](https://keras.io/) to define and train our neural network
> models.
> Keras is a machine learning framework with ease of use as one of its main features.

> > It is part of the tensorflow python package and can be imported using `from tensorflow import keras`.

Reviewer: The way to import keras is mentioned here in the call-out, so it can easily be missed. Maybe repeat it before the first code block that uses Keras

>
> Keras includes functions, classes and definitions to define deep learning models, cost functions and
> optimizers (optimizers are used to train a model).
{:.callout}

## 4. Choose a cost function and metrics
Having selected the target enoding, we need to select an appropriate loss
function that we will use during training.
This loss function tells the training algorithm how wrong, or how 'far away' from the true
value the predicted value is.

For the one hot encoding we will use the Categorical Crossentropy loss.
This is a measure for how close the distribution of the three neural network outputs corresponds
to the distribution of the three values in the one hot encoding.
It is lower if the distributions are more similar.

> In keras this is implemented in the `keras.losses.CategoricalCrossentropy` class.

Reviewer: Refer to the documentation of Keras here? To show that there's different types of loss for different types of problems.

Reviewer: You could also add here that you would want additional metrics to measure your performance such as accuracy. Or keep that for the next episode?

## 5. Choose a pretrained model or start building architecture from scratch
Now we will build a neural network from scratch, and although this sounds like
a daunting task, with Keras it is actually surprisingly straightforward.

> With keras you compose a neural network by creating layers and linking them

Reviewer: ```suggestion
With Keras you compose a neural network by creating layers and linking them
```

together. For now we will only use one type of layer called a fully connected
or Dense layer. In keras this is defined by the `keras.layers.Dense` class.

> A dense layer has a number of neurons, which is a parameter when you create the layer.

Reviewer: ```suggestion
A dense layer has a number of neurons, which is a parameter you can choose when you create the layer.
```

The hidden layer in the image in the introduction of this episode is a Dense layer.

The input in Keras also gets special treatment, Keras autmatically calculates the number of inputs
and outputs a layer needs and therefore how many edges need to be created.
This means we need to let Keras now how big our input is going to be.
We do this by instantiating a `keras.Input` class and tell it how big our input is.

~~~
inputs = keras.Input(shape=training_data.shape[1])
~~~
{:.language-python}

We store a reference to this input class in a variable so we can pass it to the creation of
our hidden layer.
Creating the hidden layer can then be done as follows:
~~~
hidden_layer = keras.layers.Dense(10, activation="relu")(inputs)
~~~
{:.language-python}

The instantiation here has 2 parameters and a seemlingly strange combination of parenthenses, so
let's take a closer look.
The first parameter `10` is the number of neurons we want in this layer, this is one of the
hyperparameters of our system and needs to be chosen carefully. We will get back to this in the section
on hyperparameter tuning.

> The second parameter is the activation function to use, here we choose relu which is 0 for values
> 0 and below and the same as the input for values above 0.

Reviewer: ```suggestion
The second parameter is the activation function to use, here we choose relu which is 0 for inputs that are
0 and below and the identity function (returning the same value) for inputs above 0.
```
And maybe add an image here?
Do we need to explain here why we need a non-linear activation function?

Now we create another layer that will be our output layer.
Again we use a Dense layer and so the call is very similar to the previous one.
~~~
output_layer = keras.layers.Dense(3, activation="softmax")(hidden_layer)
~~~
{:.language-python}
Because we chose the one hot encoding, we use `3` neurons for the output layer.
This works well in combination with the `softmax` activation function and the
Categorical Crossentropy loss we chose earlier.

The softmax activation ensures that the three output neurons produce values in the range
(0, 1) and the sum to 1.
We can interpret this as a kind of 'probability' that the sample belongs to a certain
species.
The Categorical Crossentropy loss then works well with comparing these probabilities
with 'true' probabilities that we generated using the one hot encoding.

Now that we have defined the layers of our neural network we can combine them into
a keras model which facilitates training the network.
~~~
model = keras.Model(inputs=inputs, outputs=output_layer)
model.summary()
~~~
{:.language-python}

The model summary here can show you some information about the neural network we have defined.

> ## Create the neural network
>

> > Using the code snippets above define a keras model with 1 hidden layer with
> > 10 neurons and an output layer with 3 neurons.

Reviewer: ```suggestion
> With the code snippets above, we defined a keras model with 1 hidden layer with
> 10 neurons and an output layer with 3 neurons.
```

>
> > ## Solution
> > ~~~
> > inputs = keras.Input(shape=input_data.shape[1])
> > hidden_layer = keras.layers.Dense(10, activation="relu")(inputs)
> > output_layer = keras.layers.Dense(3, activation="softmax")(hidden_layer)
> >
> > model = keras.Model(inputs=inputs, outputs=output_layer)
> > model.summary()
> > ~~~
> > {:.language-python}
> >
> > ~~~
> Model: "functional_1"
> > _________________________________________________________________
> > Layer (type)                 Output Shape              Param #
> > =================================================================
> > input_1 (InputLayer)         [(None, 4)]               0
> > _________________________________________________________________
> > dense (Dense)                (None, 10)                50
> > _________________________________________________________________
> > dense_1 (Dense)              (None, 3)                 33
> > =================================================================
> > Total params: 83
> > Trainable params: 83
> > Non-trainable params: 0
> > _________________________________________________________________
> > ~~~
> > {:.output}
> >
> > The model has 83 trainable parameters.

> > > If you increase the number of neurons in the hidden layer the number of
> > > trainable parameters increases and decreases if you decrease the number

Reviewer: ```suggestion
> > If you increase the number of neurons in the hidden layer the number of
> > trainable parameters in both the hidden and output layer increases or decreases accordingly
```

> > of neurons.
> {:.solution}
{:.challenge}

## 6. Train model
We are now ready to train the model.
Training the model requires us to make a several more choices, this time we need to
choose which optimizer to use and if this optimizer has parameters what values
to use for those.

> Furthermore the training we need to specify how many times to show the training samples.

Reviewer: ```suggestion
Furthermore, we need to specify how many times to show the training samples to the optimizer.
```

> Once more keras gives us plenty of choices all have their own pro's and cons,

Reviewer: ```suggestion
Once more, Keras gives us plenty of choices which all have their own pro's and cons,
```

it is good to play with them a little bit to see how they change and influence the
model training.
> ## Change some of the hyperparameters
>
> Change one of the hyperparameters of the neural network.
> You can for instance change the model by adding or removing layers,
> changing the activation function of the hidden layer.
> Or you can change the training, for instance by adding a learning
> rate throught the `learning_rate` parameter of the Adam optimizer.
>
> * How does changing these hyperparameters impact the training
> * What would be a good strategy to find good hyperparameter values?
{:.challenge}

## 8. Measuring Performance
Normally we would now try to assess the quality of the trained model by testing it
on some data that was not used for training.
This is often done by keeping a portion of the dataset seperate from the training data,
when using that to measure performance such a set is called a test set.
In this episode we did not prepare such a test set, however, and this will be covered
in later epidsodes.

## 9. 'predict'
Now that we have a training neural network, we can use it to predict new samples
of penguin using the `predict` function.

> This will use the network to predict the outputs we trained.

Reviewer: I think it would be good to leave out a test set at the beginning to do prediction on and a final test of our model. The next episode can then add the step of also holding out a validation set for hyperparameter tuning.

~~~
# Select the first penguin from the dataset as an example
penguin_data = training_data[0:1]

# use predict to predict the
prediction = model.predict(penguin_data)
print(prediction)
~~~
{:.language-python}

Remember that the output of the network uses the `softmax` activation function and has three
outputs, one for each species.
Therefore, we need to transform this output to one penguin species.
We can do this by taking the highest valued output and converting that to the
corresponding species.

We can do this using numpy `argmax`, which gives is the index of the highest output.
This number we can use with the categories in the species column to get the correct species
like so:
~~~
import numpy as np
species_nr = np.argmax(prediction, axis=1)

> penguin_species = penguins['species'].cat.categories[species_nr]

Reviewer: This line gives an error for me: 
```
AttributeError: Can only use .cat accessor with a 'category' dtype
```
I would do:
`predicted_species = target.columns[species_nr]`

print(penguin_species)
~~~
{:.language-python}

> 

Reviewer: As kind of a preperation for the next episode, we could do an exercise here asking to compute and compare the accuracy on the test set and the accuracy on the train set This would also test their understanding of the result of the predict method, and a refresher on prediction metrics such as accuracy.

- "Describe a fully connected layer"
- "Implement a fully connected layer with Keras"
- "Use Keras to train a small fully connected network on prepared data"
- "Plot the loss curve of the training process"
- "Use a confusion matrix to measure the trained networks' performance on a test set"
keypoints:
- The deep learning workflow is a useful tool to structure your approach, it helps to make sure you do not forget any important steps.
- Exploring the data is an important step to familiarize yourself with the problem and to help you determine the relavent inputs and outputs.
- One-hot-encoding is a preprocessing step to prepare labels for classification in Keras.
- A fully connected layer is a layer which has connections to all neurons in the previous and subsequent layers.
- keras.layers.Dense is an implement of a fully connected layer, you can set the number of neurons in the layer and the activation function used.
- To train a neural network with Keras we need to first define the network using layers and the Model class. Then we can train it using the model.fit function.
- Plotting the loss curve can be used to identify and troubleshoot the training process.

> - The loss curve does not provide any information on how well a network performs in a real setting.

Reviewer: ```suggestion
- The loss curve on the train set does not provide any information on how well a network performs in a real setting.
```

> {:.solution}
{:.challenge}

## 7. Tune hyperparameters
As we discussed before the design and training of a neural network comes with
many hyper parameter choices.
We will go into more depth of these hyperparameters in later episodes.
For now it is important to realize that the parameters we chose were
somewhat arbitrary and more careful consideration needs to be taken to
pick hyperparameter values.

## 8. Measuring Performance
Now that we have a trained neural network it is important to assess how well it performs.
We want to know how well it will perform in a realistic prediction scenario, measuring
performance will also come back when tuning the hyperparameters.

We have created a test set during the data preparation stage which we will use
now to create a confusion matrix.

### Predict the species of the test set
The first step here is to use the neural network to predict the species of the test set
using the `predict` function. This will return a `numpy` matrix, which I like to convert
to a pandas dataframe to easily see the labels.
~~~
y_pred = model.predict(X_test)
prediction = pd.DataFrame(y_pred, columns=target.columns)

> prediction

Reviewer: Missing output box here

~~~
{:.language-python}

Remember that the output of the network uses the `softmax` activation function and has three
outputs, one for each species. This dataframe shows this nicely.

We now need to transform this output to one penguin species per sample.
We can do this by looking for the index of highest valued output and converting that
to the corresponding species.
Pandas dataframes have the `idxmax` function, which will do exactly that.

~~~
predicted_species = prediction.idxmax(axis="columns")

> predicted_species

Reviewer: Also missing output box here

~~~
{:.language-python}

### Confusion matrix
With the predicted species we can now create a confusion matrix and display it
using seaborn.
To create a confusion matrix we will use another convenient function from sklearn
called `confusion_matrix`.
This function takes as a first parameter the true labels of the test set.
We can get these by using the `idxmax` method on the y_test dataframe.
The second parameter is the predicted labels which we did above.

~~~
from sklearn.metrics import confusion_matrix

true_species = y_test.idxmax(axis="columns")

matrix = confusion_matrix(true_species, predicted_species)
print(matrix)
~~~
{:.language-python}

Unfortunately, this matrix is kinda hard to read. Its not clear which column and which row
corresponds to which species.
So let's convert it to a pandas dataframe with its index and columns set to the species
as follows:

~~~
# Convert to a pandas dataframe
confusion_df = pd.DataFrame(matrix, index=y_test.columns.values, columns=y_test.columns.values)

# Set the names of the x and y axis, this helps with the readability of the heatmap.
confusion_df.index.name = 'True Label'
confusion_df.columns.name = 'Predicted Label'

> sns.heatmap(confusion_df, annot=True)

Reviewer: I think this line should be removed, it's duplicate with the next code block.

~~~
{:.language-python}

We can then use the `heatmap` function from seaborn to create a nice visualization of
the confusion matrix.
The `annot=True` parameter here will put the numbers from the confusion matrix in
the heatmap.

~~~
sns.heatmap(confusion_df, annot=True)
~~~
{:.language-python}

> 

Reviewer: I would show the output image of the heatmap here, instead of in the exercise (as you give the code already above the exercise). The exercise is then just about interpreting the results.
